arctic_training/data/dpo_factory.py

```diff
@@ -0,0 +1,347 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict
+from typing import List
+from typing import Tuple
+from typing import Union
+
+import torch
+from torch.utils.data import DataLoader
+from torch.utils.data import RandomSampler
+from transformers import BatchEncoding
+from transformers import PreTrainedTokenizerBase
+
+from arctic_training.config.data import DataConfig
+from arctic_training.data.factory import DataFactory
+from arctic_training.data.sft_factory import pad
+from arctic_training.data.utils import DatasetType
+
+IGNORE_INDEX = -100
+
+
+class DPODataConfig(DataConfig):
+    max_length: int = 8192
+    """ Maximum length of the input sequence. """
+
+    max_prompt_length: int = 4096
+    """ Maximum prompt length of the input sequence. """
+
+    mask_inputs: bool = True
+    """ Whether to mask the input sequence. """
+
+    always_max_length: bool = False
+    """
+    If this is turned on, each batch will be filled up to the max length by
+    appending samples until the total length matches the max length. It might
+    cause the last sample to be truncated.
+    """
+
+    dpo_prompt_truncation_mode: str = "keep_start"
+
+
+def _adjust_prompt_length(
+    prompt_token,
+    chosen_token,
+    rejected_token,
+) -> None:
+    c_len = len(chosen_token["prompt_input_ids"])
+    r_len = len(rejected_token["prompt_input_ids"])
+    min_len = min(c_len, r_len)
+
+    for k, v in prompt_token.items():
+        prompt_token[k] = v[:min_len]
+
+    num_diff_tokens = sum(
+        [
+            a != b
+            for a, b in zip(
+                chosen_token["prompt_input_ids"], rejected_token["prompt_input_ids"]
+            )
+        ]
+    )
+    num_diff_len = abs(c_len - r_len)
+    if num_diff_tokens > 1 or num_diff_len > 1:
+        raise ValueError(
+            "Chosen and rejected prompt_input_ids might only differ on the last token"
+            " due to tokenizer merge ops."
+        )
+
+
+def add_bos_token_if_needed(
+    bos_token_id: Union[None, int],
+    tokens: Dict[str, List[int]],
+) -> Dict[str, List[int]]:
+    len_input_ids = len(tokens["prompt_input_ids"])
+    if bos_token_id is not None and (
+        len_input_ids == 0 or bos_token_id != tokens["prompt_input_ids"][0]
+    ):
+        tokens["prompt_input_ids"].insert(0, bos_token_id)
+        tokens["prompt_attention_mask"].insert(0, 1)
+    return tokens
+
+
+def _build_sequence_tokens(tokens, prefix: str) -> Dict:
```

Comment:
```suggestion
def _build_sequence_tokens(tokens: Dict[str, List[int]], prefix: str) -> Dict[str, List[int]]:
```
Better define type hints